File: LeetCode/2. Add Two Numbers.py
# https://leetcode.com/problems/add-two-numbers/description/
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(
        self, head1: Optional[ListNode], head2: Optional[ListNode]
    ) -> Optional[ListNode]:
        current1: ListNode | None = head1
        current2: ListNode | None = head2

        head: ListNode | None = None
        tail: ListNode | None = None

        rest = 0

        while current1 or current2:
            value1 = getattr(current1, "val", 0)
            value2 = getattr(current2, "val", 0)

            sum = value1 + value2 + rest

            node = ListNode(sum % 10)

            rest = 1 if sum >= 10 else 0

            if head is None or tail is None:
                head = node
                tail = node
            else:
                tail.next = node
                tail = node

            current1 = getattr(current1, "next", None)
            current2 = getattr(current2, "next", None)

        if rest and tail:
            node = ListNode(rest)
            tail.next = node

        return head

File: LeetCode/2. Add Two Numbers.js
// https://leetcode.com/problems/add-two-numbers/description/

const addTwoNumbers = function (head1, head2) {
  let current1 = head1;
  let current2 = head2;

  let head = null;

  let tail = null;

  let rest = 0;

  while (current1 || current2) {
    const value1 = current1?.val || 0;
    const value2 = current2?.val || 0;

    const sum = value1 + value2 + rest;

    rest = sum >= 10 ? 1 : 0;

    const node = new ListNode(sum % 10);

    if (!head) {
      head = node;
      tail = node;
    } else {
      tail.next = node;
      tail = node;
    }

    current1 = current1?.next;
    current2 = current2?.next;
  }

  if (rest) {
    const node = new ListNode(rest);

    tail.next = node;
  }

  return head;
};
